src/runtime-mpi.c

```diff
@@ -0,0 +1,293 @@
+        return 4;
+    }
+
+    if (MPI_SUCCESS != MPI_Comm_size(SHMEM_RUNTIME_WORLD, &size)) {
+        return 5;
+    }
+
+    if (size > 1) { 
+
+    	max_name_len = 128;
+        max_key_len = 128;
+        max_val_len = 128;
+
+        //WARN: THIS ASSUMES THAT YOU CAN ONLY HAVE 10 KEY VALUE PAIRS
+        kv_store_all = (char**) malloc(sizeof(char*) * MAX_KV_COUNT * 2); 
+        if(NULL == kv_store_all) return 6; //rename
+        kv_store_me = (char**) malloc(sizeof(char*) * MAX_KV_COUNT * 2);
+        if(NULL == kv_store_me) return 7; //rename
+
+    }
+    else {
+        /* Use a local KVS for singleton runs */
+        max_key_len = SINGLETON_KEY_LEN;
+        max_val_len = SINGLETON_VAL_LEN;
+        kvs_name = NULL;
+        max_name_len = 0;
+    }
+
+    kvs_key = (char*) malloc(max_key_len);
+    if (NULL == kvs_key) return 8;
+
+    kvs_value = (char*) malloc(max_val_len);
+    if (NULL == kvs_value) return 9;
+
+    return 0;
+}
+
+int
+shmem_runtime_fini(void)
+{
+    int finalized = 0;
+    MPI_Finalized(&finalized);
+    if(!finalized){
+        MPI_Finalize();
+    }
+    free(kv_store_all);
+    free(kv_store_me);
+
+    return 0;
+}
+
+void
+shmem_runtime_abort(int exit_code, const char msg[])
+{
+
+#ifdef HAVE___BUILTIN_TRAP
+    if (shmem_internal_params.TRAP_ON_ABORT)
+        __builtin_trap();
+#endif
+
+    if (size == 1) {
+        fprintf(stderr, "%s\n", msg);
+        exit(exit_code);
+    }
+
+    MPI_Abort(SHMEM_RUNTIME_WORLD, exit_code);
+
+    /* MPI_Abort should not return */
+    abort();
+}
+
+
+int
+shmem_runtime_get_rank(void)
+{
+    return rank;
+}
+
+
+int
+shmem_runtime_get_size(void)
+{
+    return size;
+}
+
+int
+shmem_runtime_exchange(void)
+{
+    /* Use singleton KVS for single process jobs */
+    if (size == 1){
+        return 0;
+    }
+    
+    for(int i = 0; i < length; i++)
+    {
+        *(kv_store_all + i) = (char*)malloc(max_val_len * sizeof(char) * size);
+
+        if (MPI_SUCCESS != MPI_Allgather(*(kv_store_me+i), max_val_len, MPI_CHAR, *(kv_store_all + i), max_val_len, MPI_CHAR, SHMEM_RUNTIME_WORLD)) {
+            return 2;   
+        }
+        if (MPI_SUCCESS != MPI_Barrier(SHMEM_RUNTIME_WORLD)) {
+            return 6;
+        }
+    }
+    return 0;
+}
+
+
+int 
+shmem_runtime_put(char *key, void *value, size_t valuelen)
+{
+    snprintf(kvs_key, max_key_len, "shmem-%lu-%s", (long unsigned) rank, key);
+    if (0 != encode(value, valuelen, kvs_value, max_val_len)) {
+        return 1;
+    }
+
+    if (size == 1) {
+        singleton_kvs_t *e = malloc(sizeof(singleton_kvs_t));
+        if (e == NULL) return 3;
+        strncpy(e->key, kvs_key, max_key_len);
+        strncpy(e->val, kvs_value, max_val_len);
+        HASH_ADD_STR(singleton_kvs, key, e);
+    } else {
+        if(length < 20){
+            *(kv_store_me + length) = key;
+            length++;
+            *(kv_store_me + length) = kvs_value;
```

Comment:
The key and value need to be copied here or the data might be corrupted/overwritten before the call to gather.